tools/buildgen/generate_projects.sh

```diff
@@ -18,6 +18,9 @@ set -e
 
 export TEST=${TEST:-false}
 
+# Upgrade Python's YAML library
+python3 -m pip install --upgrade --ignore-installed PyYAML --user
```

Comment:
@lidizheng 
Just noticed this sneaky line.  The problem is it's fragile since it always downloads pyyaml (even if a perfectly fine version is already installed) and will lead to test flakiness sooner or later.
- sanity tests are supposed to be hermetic (to avoid flakiness), but like this they download pyyaml on each run, risking flakes (experience shows things like this eventually fail, expecially if the version to install is not pinned).
- I see a pyyaml installation in the logs every time I run generate_projects.sh locally.
- the way things currently work is inconsistent, since it installs the pyyaml packages if it's missing, but doesn't install other python packages require for "regenerate" (such a mako). This is counterintuitive.

Let's find a better solution in a followup PR.

Reply:
Do you prefer to:

1. Have the script refresh all generation dependencies;
2. Tweak sanity test image to have newer "yaml", whether from `apt` or `pip`.

Or we can do both to ensure the consistent output from Kokoro and manual runs.

Reply:
- The script should now download/refresh anything unless absolutely needed (if you have a new enough version of pyyaml, it shouldn't download again).
- under sanity docker image, the script should not install anything - everything needed should be pre-installed in the dockerimage (= which is much less flaky that downloading a pip package which can fail or there could be a new version of the pip package that breaks our stuff for some reason).
- if you want, you can make the script detect whether a new-enough version of pyyaml is installed and install it automatically if not, but I think perhaps the safest way is to just test whether all the python dependencies are installed and fails with a clear message that tells you what to install. Installing the right pip packages is easy, it's a one time jobs and `generate_projects.sh` is not something you use that often anyway. I think scripts that try to do too much magic for you are actually harmful (because they need maintenance and often break).